nameresolution/mdns/mdns_test.go

```diff
@@ -162,6 +239,223 @@ func TestResolverMultipleInstances(t *testing.T) {
+// series as they rely on a shared mDNS server on the host
+// machine.
+func TestResolverConcurrency(t *testing.T) {
+	tt := []struct {
+		name string
+		test func(t *testing.T)
+	}{
+		{
+			name: "ResolverConcurrencyNotFound",
+			test: ResolverConcurrencyNotFound,
+		},
+		{
+			name: "ResolverConcurrencyFound",
+			test: ResolverConcurrencyFound,
+		},
+		{
+			name: "ResolverConcurrencySubscriberClear",
+			test: ResolverConcurrencySubsriberClear,
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, tc.test)
+	}
+}
+
+func ResolverConcurrencySubsriberClear(t *testing.T) {
+	// arrange
+	resolver := NewResolver(logger.NewLogger("test"))
+	defer resolver.Close()
+	md := nr.Metadata{Properties: map[string]string{
+		nr.MDNSInstanceName:    "testAppID",
+		nr.MDNSInstanceAddress: localhost,
+		nr.MDNSInstancePort:    "1234",
+	}}
+
+	// act
+	err := resolver.Init(md)
+	require.NoError(t, err)
+
+	request := nr.ResolveRequest{ID: "testAppID"}
+
+	var wg sync.WaitGroup
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			pt, err := resolver.ResolveID(request)
+			require.NoError(t, err)
+			require.Equal(t, fmt.Sprintf("%s:1234", localhost), pt)
+		}()
+	}
+
+	wg.Wait()
+
+	// Wait long enough for the background clear to occur.
+	time.Sleep(3 * time.Second)
+	require.Equal(t, 0, len(resolver.subs))
+}
+
+// WARN: This is deliberately not a test function.
+// This test case must be run in serial and is executed
+// by the TestResolverConcurrency test function.
+func ResolverConcurrencyFound(t *testing.T) {
+	// arrange
+	resolver := NewResolver(logger.NewLogger("test"))
+	defer resolver.Close()
+
+	// register instance A
+	appAID := "A"
+	appAName := "testAppA"
+	appAAddress := localhost
+	appAPort := "1234"
+	appAPQDN := fmt.Sprintf("%s:%s", appAAddress, appAPort)
+
+	appA := nr.Metadata{Properties: map[string]string{
+		nr.MDNSInstanceName:    appAName,
+		nr.MDNSInstanceAddress: appAAddress,
+		nr.MDNSInstancePort:    appAPort,
+		nr.MDNSInstanceID:      appAID,
+	}}
+	err1 := resolver.Init(appA)
+	require.NoError(t, err1)
+
+	// register instance B
+	appBID := "B"
+	appBName := "testAppB"
+	appBAddress := localhost
+	appBPort := "5678"
+	appBBPQDN := fmt.Sprintf("%s:%s", appBAddress, appBPort)
+
+	appB := nr.Metadata{Properties: map[string]string{
+		nr.MDNSInstanceName:    appBName,
+		nr.MDNSInstanceAddress: appBAddress,
+		nr.MDNSInstancePort:    appBPort,
+		nr.MDNSInstanceID:      appBID,
+	}}
+	err2 := resolver.Init(appB)
+	require.NoError(t, err2)
+
+	// register instance C
+	appCID := "C"
+	appCName := "testAppC"
+	appCAddress := localhost
+	appCPort := "3456"
+	appCBPQDN := fmt.Sprintf("%s:%s", appCAddress, appCPort)
+
+	appC := nr.Metadata{Properties: map[string]string{
+		nr.MDNSInstanceName:    appCName,
+		nr.MDNSInstanceAddress: appCAddress,
+		nr.MDNSInstancePort:    appCPort,
+		nr.MDNSInstanceID:      appCID,
+	}}
+	err3 := resolver.Init(appC)
+	require.NoError(t, err3)
+
+	// act...
```

Comment:
In this test case, I can see that Init is called on the same resolver 3 times. Is this to facilitate the concurrency test? 
Because my understanding is when daprd starts, NewResolver is called once and Init is called when initNameResolution() is called from the runtime.go file.

Reply:
yes this is not the behaviour used by the runtime but just the tests.